### DISCUSSION TITLE
Pasta equivalent to macvlan

### DISCUSSION THREAD

**Discussion by skyblaster** [containers/podman#22709] [Q&A]
With Pasta now the default for rootless networking, I'm finally motivated to convert all of my containers to rootless.
I've been experimenting but have been left a little confused. Specifically what the purpose of the `-a` and `-o` options are for.

If I run the following command, my container is only reachable via the host IP address:
```
podman run -d --network=pasta:--ipv4-only,-a,10.10.0.20,-n,24,-g,10.10.0.1,--map-gw,-o,10.10.0.20 -p 3000:3000/tcp docker.io/adguard/adguardhome

$ curl http://10.10.0.20:3000
curl: (7) Failed to connect to 10.10.0.20 port 3000 after 3089 ms: Couldn't connect to server

$ curl http://10.10.0.227:3000
<a href="/install.html">Found</a>.
```

If I remove `-p 3000:3000/tcp` from the command, then the container is not reachable at all. With or without `--map-gw` makes no difference.

Now, I wasn't expecting full macvlan capability, but based on the switches I was at least hoping.

So, having read the [shortcomings of rootless podman](https://github.com/containers/podman/blob/main/rootless.md), is it possible (without involving NAT or a proxy) to serve up the same port from two separate rootless containers on the same host similar to macvlan? If yes, are there examples one can follow?

Thanks!

**Comment by sbrivio-rh** [comment]
> With Pasta now the default for rootless networking, I'm finally motivated to convert all of my containers to rootless. I've been experimenting but have been left a little confused. Specifically what the purpose of the `-a` and `-o` options are for.

`-a` just assigns a given address to the interface *in the container*: pasta doesn't run as root, so it doesn't and can't affect any network configuration on the host. `-o` is about how traffic from the container will look like to the outside world (it just picks source address/interface, essentially).

> If I run the following command, my container is only reachable via the host IP address:
> 
> ```
> podman run -d --network=pasta:--ipv4-only,-a,10.10.0.20,-n,24,-g,10.10.0.1,--map-gw,-o,10.10.0.20 -p 3000:3000/tcp docker.io/adguard/adguardhome
> 
> $ curl http://10.10.0.20:3000
> curl: (7) Failed to connect to 10.10.0.20 port 3000 after 3089 ms: Couldn't connect to server
> 
> $ curl http://10.10.0.227:3000
> <a href="/install.html">Found</a>.
> ```
> 
> If I remove `-p 3000:3000/tcp` from the command, then the container is not reachable at all. With or without `--map-gw` makes no difference.

...that's expected: Podman doesn't want pasta accidentally "leaking" container ports to the host, so you have to forward ports explicitly. pasta, by itself, would otherwise give you `-t auto` (periodically scans ports bound in the container and forwards them). But you can [override](https://github.com/containers/common/blob/main/libnetwork/pasta/pasta_linux.go#L153) that if you want, with `-t,auto`.

Still, at the moment, you can't use that feature together with a specific address, that is, you can't say `-t 10,10.0.20/auto`. We're working on a more flexible port forwarding model based on a full-fledged [flow table](https://archives.passt.top/passt-dev/20240514010337.1104606-1-david@gibson.dropbear.id.au/), but there's still some work to do.

After that, we'll probably be able to support something like this: https://bugs.passt.top/show_bug.cgi?id=48. It might still look a bit more convoluted than macvlan, but at that point we should have equivalent functionality implemented in userspace.

> Now, I wasn't expecting full macvlan capability, but based on the switches I was at least hoping.
> 
> So, having read the [shortcomings of rootless podman](https://github.com/containers/podman/blob/main/rootless.md), is it possible (without involving NAT or a proxy) to serve up the same port from two separate rootless containers on the same host similar to macvlan? If yes, are there examples one can follow?

You can bind specific ports to two different addresses. With stand-alone pasta (*not* a container, just as a network configuration example):

```
$ pasta -t 192.0.2.1/3000
# nc -l 3000
```

and another network namespace can have, at the same time,

```
$ pasta -t 192.0.2.2/3000
# nc -l 3000
```

if both 192.0.2.1 and 192.0.2.2 are assigned the host.

Note that the wish from https://bugs.passt.top/show_bug.cgi?id=48 goes beyond that: with IPv6 and non-local binds enabled, containers could have their own address sharing a prefix with the host, and we wouldn't need to configure those specific addresses on the host at all, so that would be some kind of zero-configuration thing.
